python/dgl/distributed/sparse_emb.py

```diff
@@ -58,148 +59,147 @@ class DistEmbedding:
+        self._part_policy = part_policy
+
+    def __call__(self, idx, device=th.device('cpu')):
+        """
+        node_ids : th.tensor
+            Index of the embeddings to collect.
+        device : th.device
+            Target device to put the collected embeddings.
+
+        Returns
+        -------
+        Tensor
+            The requested node embeddings
+        """
         idx = utils.toindex(idx).tousertensor()
-        emb = self._tensor[idx]
+        emb = self._tensor[idx].to(device, non_blocking=True)
         if F.is_recording():
             emb = F.attach_grad(emb)
-            self._trace.append((idx, emb))
+            self._trace.append((idx.to(device, non_blocking=True), emb))
         return emb
 
     def reset_trace(self):
         '''Reset the traced data.
         '''
         self._trace = []
 
-class SparseAdagradUDF:
-    ''' The UDF to update the embeddings with sparse Adagrad.
+    @property
+    def part_policy(self):
+        """Return the partition policy
 
-    Parameters
-    ----------
-    lr : float
-        The learning rate.
-    '''
-    def __init__(self, lr):
-        self._lr = lr
+        Returns
+        -------
+        PartitionPolicy
+            partition policy
+        """
+        return self._part_policy
 
-    def __call__(self, data_store, name, indices, data):
-        ''' Update the embeddings with sparse Adagrad.
+    @property
+    def rank(self):
+        """Return rank of current process.
 
-        This function runs on the KVStore server. It updates the gradients by scaling them
-        according to the state sum.
+        Returns
+        -------
+        int
+            The rank of current process.
+        """
+        return self._rank
 
-        Parameters
-        ----------
-        data_store : dict of data
-            all data in the kvstore.
-        name : str
-            data name
-        indices : tensor
-            the indices in the local tensor.
-        data : tensor (mx.ndarray or torch.tensor)
-            a tensor with the same row size of id
-        '''
-        grad_indices = indices
-        grad_values = data
-        embs = data_store[name]
-        state_sum = data_store[name + "_sum"]
-        with F.no_grad():
-            grad_sum = F.mean(grad_values * grad_values, 1)
-            F.index_add_inplace(state_sum, grad_indices, grad_sum)
-            std = state_sum[grad_indices]  # _sparse_mask
-            std_values = F.unsqueeze((F.sqrt(std) + 1e-10), 1)
-            F.index_add_inplace(embs, grad_indices, grad_values / std_values * (-self._lr))
+    @property
+    def world_size(self):
+        """Return world size of the pytorch distributed training env.
 
-def _init_state(shape, dtype):
-    return F.zeros(shape, dtype, F.cpu())
+        Returns
+        -------
+        int
+            The world size of the pytorch distributed training env.
+        """
+        return self._world_size
 
-class SparseAdagrad:
```

Comment:
can we keep this async optimizer? we should make a performance comparison between the sync version and async version.